fix: immutable resource status

waciumawanjohi opened the pull request:
<!--
Thanks for submitting a pull request to Cartographer!

Also check the [PR guidelines] if you haven't already!
-->

[PR requirements]: https://github.com/vmware-tanzu/cartographer/blob/main/CONTRIBUTING.md#commit-message-and-pr-guidelines

## Changes proposed by this PR

<!--
Add the story that is resolved or updated
`closes`/`fixes` or `updates` are valid here 
-->
Resolves #1120

Also resolves an issue where `lifecycle:tekton` resources would remain in a healthy state, even if the most recent object was in an unknown or unhealthy state.

<!--
Summarize your changes. Please include reasoning and key decisions to help the reviewer
understand the changes.
-->


## Release Note

<!--
Your PR title will be directly included in the release notes when it ships in
the next release. It should briefly describe the PR in [imperative
mood]. Please refrain from adding prefixes like 'feature:', and don't include a
period at the end.

Within this section you may supply a list of extra information to include in
the release notes in addition to the pull request title.

Example title: Add CreatorName field in Workload spec

Example notes:

* Operators can label stamped objects with the Workload's creator
* Creators can be contacted when interdepartmental issues arise 

If there are no additional notes necessary you may remove this entire section.
-->
[imperative mood]: https://chris.beams.io/posts/git-commit/#imperative

## PR Checklist

Note: Please do not remove items. Mark items as done `[x]` or use ~strikethrough~ if you believe they are not relevant

- [x] Linked to a relevant issue. Eg: `Fixes #123` or `Updates #123`
- [ ] Removed non-atomic or `wip` commits
- ~[ ] Filled in the [Release Note](#Release-Note) section above~
- ~[ ] Modified the docs to match changes~

waciumawanjohi committed:
Demo error

TODO: delete addition of itResultsInAHealthyWorkload()

waciumawanjohi committed:
Refactor

waciumawanjohi committed:
Reset test to proper state

waciumawanjohi committed:
More refactoring

waciumawanjohi committed:
Improve message when no immutable object is healthy

waciumawanjohi committed:
Move tekton health check into templates

waciumawanjohi committed:
Test status when no tekton objects found

waciumawanjohi committed:
Improve immutable test

waciumawanjohi committed:
Tighten expectations

waciumawanjohi committed:
Test fix

squeedee commented on tests/integration/supplychain/workload_reconciler_test.go:
```diff
@@ -683,23 +687,50 @@ var _ = Describe("WorkloadReconciler", func() {
+							"SupplyChainReady": MatchFields(IgnoreExtras, Fields{
+								"Status": Equal(metav1.ConditionTrue),
+							}),
+							"ResourcesSubmitted": MatchFields(IgnoreExtras, Fields{
```
I'm concerned by the confusion this might create in the top level conditions:
```
  Type: ResourcesSubmitted
  Status: False
  Reason: NoHealthyStampedObjectsFound
```

It reads to me like "nothing across the supply chain is healthy" (I can logically deduce that this is not what it means, but I'd still be forced to think about it)

Perhaps: 
```
  Type: ResourcesSubmitted
  Status: False
  Reason: NoHealthyStampedImmutableObjectFound
```

waciumawanjohi committed:
Clarify error message

When propagated to the top level workload condition, this message should clarify what is unhealthy
